torch/_functorch/aot_autograd.py

```diff
@@ -1549,111 +1707,82 @@ def compiled_function(*args):
-                        # Functionalization can't easily tell us if an input had BOTH its metadata actual data mutated.
-                        # So we check if metadata needs to be mutated here manually.
+                    if meta.mutation_type == MutationType.data_and_metadata:
                         original_inpt.as_strided_(updated_inpt.size(), updated_inpt.stride(), updated_inpt.storage_offset())
+                    else:
+                        assert meta.mutation_type == MutationType.data
                     original_inpt.copy_(updated_inpt)
         else:
-            fw_outs = outs
+            fw_outs = all_outs
 
         # Step 4: Manually regenerate any outputs that are aliased to inputs, instead of
         # compiling them.
-        if CompiledFunction.num_outputs_aliased_to_inputs > 0 or CompiledFunction.num_outputs_aliased_to_intermediates > 0:
-            assert CompiledFunction.num_outputs_aliased_to_inputs + len(fw_outs) == \
-                len(CompiledFunction.fw_metadata.aliased_output_info)
+        if CompiledFunction.num_outputs_aliased > 0:
+            assert len(fw_outs) == len(CompiledFunction.fw_metadata.output_info)
+            # The compiled forward also returned intermediate bases. We don't want to return them to the user.
+            if CompiledFunction.fw_metadata.num_intermediate_bases > 0:
+                fw_outs_no_intermediate_bases = fw_outs[:-CompiledFunction.fw_metadata.num_intermediate_bases]
+            else:
+                fw_outs_no_intermediate_bases = fw_outs
+
             fw_outs_including_aliases = []
-            for aliased_out_metadata in CompiledFunction.fw_metadata.aliased_output_info:
-                if aliased_out_metadata.output_type == OutputType.non_alias:
-                    fw_outs_including_aliases.append(fw_outs[aliased_out_metadata.base_idx])
-                else:
-                    if aliased_out_metadata.output_type == OutputType.alias_of_input:
-                        aliased_base_tensor = args[aliased_out_metadata.base_idx]
+            for o, info in zip(fw_outs_no_intermediate_bases, CompiledFunction.fw_metadata.output_info):
+                if info.output_type == OutputType.non_alias:
+                    fw_outs_including_aliases.append(o)
+                    continue
+                assert isinstance(o, TensorAlias)
+                o_ = o.alias
+                o_grad = o.requires_grad
+                if info.output_type == OutputType.alias_of_input:
+                    aliased_base_tensor = args[info.base_idx]
+                    if info.is_input_tensor:
+                        # Special case for when an output *is* an input (just return the input, don't generate a view)
+                        # Note: we can't just detect this at runtime, because of this case:
+                        # def f(x):
+                        #     x.mul_(2)
+                        #     return x
```

Comment:
I don't understand this comment

Reply:
This is just in reference to a bug I noticed when I was refactoring. Yep, it's better for performance to pre-compute which outputs **are** inputs - but it's also required. We don't actually know if an output of `autograd.Function.apply()` **is** an input at runtime.

Specifically, for the above function that mutates an input and directly returns it, we can't actually detect that an output of `autograd.Function.apply()` **is** an input at runtime
```
def wrapper(x):
    out = CompiledFunction.apply(x)
    # This doesn't work, because our functionalized graph will return `x_updated`, not `x`
    # So we **need** to precompute which outputs are the same as inputs, during tracing.
    if out is x:
        ...
```